readthedocs/search/api.py

```diff
@@ -78,43 +78,43 @@ def paginate_queryset(self, queryset, request, view=None):
         self.request = request
 
         page_size = self.get_page_size(request)
-
-        total_count = 0
-        total_pages = 1
-        if queryset:
-            total_count = queryset.total_count()
-            hits = max(1, total_count)
-            total_pages = ceil(hits / page_size)
-
         page_number = request.query_params.get(self.page_query_param, 1)
-        if page_number in self.last_page_strings:
-            page_number = total_pages
 
         original_page_number = page_number
         page_number = self._get_page_number(page_number)
+
         if page_number <= 0:
             msg = self.invalid_page_message.format(
                 page_number=original_page_number,
                 message=_("Invalid page"),
             )
             raise NotFound(msg)
 
+        start = (page_number - 1) * page_size
+        end = page_number * page_size
+
+        if queryset:
+            result = queryset[start:end].execute()
+            total_count = result.hits.total['value']
+            hits = max(1, total_count)
+            total_pages = ceil(hits / page_size)
+        else:
+            result = queryset
+            total_count = 0
+            total_pages = 1
```

Comment:
We can move this defaults to just before the if. And result should default to [].